Introduce delete_by and destroy_by methods to ActiveRecord::Relation

abhaynikam opened the pull request:
closes #35304 

Adds `delete_by` and `destroy_by` as relation methods to add missing symmetry.

@r? @dhh

abhaynikam committed:
Introduce delete_by and destroy_by methods to ActiveRecord::Relation

dhh commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,31 @@ def delete_all
+    #   Person.all.destroy_by("published_at < ?", 2.weeks.ago)
+    #   Person.where(["user_name = ?", user_name]).destroy_by(name: 'Spartacus')
+    def destroy_by(arg, *args)
+      where(arg, *args).destroy_all
```
Why not just `*args`? What's the purpose of splitting the first one out?

abhaynikam replied:
Just `*args` make sense. Fixed it. 👍

dhh commented:
I think we can pare back the testing a lot. This is basically just a delegator to where + destroy/delete_all. So we just need to test the delegation, not all the ways `where` works.

Also, should probably just spell out that this is shorthand for where().delete/destroy_all.

abhaynikam commented:
@dhh : Cutoff most of the test cases and added some basic test case to check delegation. Also, updated the comments to mention `shorthand for where().delete/destroy_all.`

dhh commented:
Looking good. Waiting for the builds to finish, but otherwise I believe good to merge.

kamipo commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+    `destroy_all` on the matched records.
+
+    Example:
+      persons = Person.all
```
Need a blank line above.

kamipo commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+    `delete_all` on the matched records.
+
+    Example:
+      persons = Person.all
```
Need a blank line above.

kamipo commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+    Example:
+      persons = Person.all
+
+      persons.destroy_by(id: [1, 2, 3])
```
Need more two space indentation.

kamipo reviewed (commented):
Broken markup https://github.com/rails/rails/blob/784ff5104be2412d8a628fa3c343f0ee56ffbe7f/activerecord/CHANGELOG.md

kamipo commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,34 @@
+      david = Person.find_by(name: 'David')
+      david.posts.delete_by("title like ?", "%Paying tribute to the web with View Source%")
+
+
```
✂️ a blank line.

kamipo commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,32 @@ def delete_all
+    # Finds and destroys all records matching the specified conditions.
+    # This is short-hand for `relation.where(condition).destroy_all`.
+    # Returns the collection of objects that were destroyed.
+
```
```suggestion
    #
```

kamipo commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,32 @@ def delete_all
+    # Finds and deletes all records matching the specified conditions.
+    # This is short-hand for `relation.where(condition).delete_all`.
+    # Returns the number of rows affected.
+
```
```suggestion
    #
```

abhaynikam commented:
@kamipo : Fixed all suggestions. Thank You. 👍

kamipo commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,32 @@ def delete_all
     end
 
+    # Finds and destroys all records matching the specified conditions.
+    # This is short-hand for `relation.where(condition).destroy_all`.
```
```suggestion
    # This is short-hand for <tt>relation.where(condition).destroy_all</tt>.
```

kamipo commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,32 @@ def delete_all
+    end
+
+    # Finds and deletes all records matching the specified conditions.
+    # This is short-hand for `relation.where(condition).delete_all`.
```
```suggestion
    # This is short-hand for <tt>relation.where(condition).delete_all</tt>.
```

kaspth commented on activerecord/lib/active_record/relation.rb:
```diff
@@ -507,6 +507,32 @@ def delete_all
+    # If no record is found, returns empty array.
+    #
+    #   Person.all.destroy_by(name: 'Spartacus', rating: 4)
+    #   Person.all.destroy_by("published_at < ?", 2.weeks.ago)
```
This doesn’t seem to match @dhhs description. I don’t think we want `.all` before a destroy_by. It should work like find_by straight on Person.

kaspth commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,35 @@
+
+        persons.destroy_by(id: [1, 2, 3])
+        persons.destroy_by(name: 'David')
+        persons.destroy_by(name: 'David', rating: 4)
```
I think the plural of person is people (at the very least Rails’ inflector says so), so let’s go with that.

abhaynikam commented:
@kaspth : Done with the changes. Please have a look.

kamipo commented on activerecord/lib/active_record/core.rb:
```diff
@@ -212,6 +212,32 @@ def find_by!(*args) # :nodoc:
+      #   Person.destroy_by(name: 'Spartacus', rating: 4)
+      #   Person.destroy_by("published_at < ?", 2.weeks.ago)
+      def destroy_by(*args)
+        where(*args).destroy_all
```
Just use `delegate` like as other query methods.

https://github.com/rails/rails/blob/ef602f8892fa4c41d7b0bb7868fc9084a9888212/activerecord/lib/active_record/querying.rb#L5-L16

abhaynikam commented:
@kaspth : Fixed all the review comments. 

cc/ @dhh @kamipo

abla00 commented on activerecord/test/cases/relations_test.rb:
```diff
@@ -1686,6 +1686,24 @@ def test_presence
+    assert_difference("Post.count", -3) { david.posts.destroy_by(body: "hello") }
+
+    destroyed = Author.destroy_by(id: david.id)
+    assert_equal [authors(:david)], destroyed
```
It's easier to read:

`assert_equal [david], destroyed`

kamipo commented:
Thanks!

adamjubert commented:
Is there any reason the name `delete_by()` was chosen as opposed to `delete_where()`? This nomenclature seems counterintuitive.

dhh commented:
To match find_by(*conditions). Also, because  I think it flows much better.
#delete_where is an ugly looking method to my aesthetics.

On Thu, Nov 21, 2019 at 9:04 AM Adam Jubert <notifications@github.com>
wrote:

> Is there any reason the name delete_by() was chosen as opposed to
> delete_where()? This nomenclature seems counterintuitive.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/35316?email_source=notifications&email_token=AAAAVNPTCCKZ2IIBURWVCW3QU25SLA5CNFSM4GYFMWVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE26IEI#issuecomment-557179921>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAVNKVXYKWRNOUPMFIIK3QU25SLANCNFSM4GYFMWVA>
> .
>

goalaleo commented:
I agree with @adamjubert . If `destroy_by` is basically just a delegator to 
```
where(*args).destroy_all
````
then what do you call the method that delegates to
```
find_by(*args).destroy
```
Can't `destroy_by` and `destroy_where` both exist?

dhh commented:
There’s no need for a singular destroy because we are not returning any records. Unlike the finder which needs different methods for singular and plural returns.

No interest in destroy_where from my side. Syntactic sugar earns its sprinkles in large parts by the fluidity of the naming. destroy_where is an ugly method name to my eyes, breaks the link to the other _by methods, and adds nothing new not already present in the existing API. 

> On Nov 26, 2019, at 02:38, Leo Kiiski <notifications@github.com> wrote:
> 
> ﻿
> I agree with @adamjubert . If destroy_by is basically just a delegator to
> 
> where(*args).destroy_all
> then what do you call the method that delegates to
> 
> find_by(*args).destroy
> Can't destroy_by and destroy_where both exist?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

goalaleo commented:
>destroy_where is an ugly method name to my eyes, breaks the link to the other _by methods, and adds nothing new not already present in the existing API.

I agree that it's not pretty. But what in your opinion is the link to the other _by methods?
If I saw the method name `destroy_by` for the first time, my intuition would tell me that it finds the first matching record, and calls `.destroy` on it. Don't all the other `_by` methods deal in singular records?

>we are not returning any records

I'm not sure what *we* refers to here. Are you talking about the `destroy_by` method? If so, then doesn't it return all destroyed records ?

westonganger commented:
There is no point to "add missing symmetry" if the methods that are added do not symmetrically match the outcome, it only symmetrically matches the method name which is ridiculous.

I vote to remove these methods. The original solution of `where().destroy_all` or `where().delete_all` is much clearer. I don't ever want to have to deal with this code in a future project.

dhh commented:
Your vote is noted. Thank you for sharing your opinion. I don't see this
changing, though. I'm very happy with the API as it currently is, and I've
used it a bunch in apps already. I'm going to bow out of the bikeshed now ✌️

On Mon, Dec 2, 2019 at 10:00 AM Weston Ganger <notifications@github.com>
wrote:

> There is no point to "add missing symmetry" if the methods that are added
> do not symmetrically match the outcome, it only symmetrically matches the
> method name which is ridiculous.
>
> I vote to remove these methods. The original solution of
> where().destroy_all or where().delete_all is much clearer. I don't ever
> want to have to deal with this code in a future project.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/rails/rails/pull/35316?email_source=notifications&email_token=AAAAVNNCKNX3Q5ZC2HHG7ELQWVEMPA5CNFSM4GYFMWVKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEFULABY#issuecomment-560508935>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAVNIVEUSKWRWISCJ4JNDQWVEMPANCNFSM4GYFMWVA>
> .
>
